### DISCUSSION TITLE
Pages is now also standalone. What does it imply (or not)?

### DISCUSSION THREAD

**Discussion by woluweb** [joomlatools/joomlatools-pages#750] [Q&A] [answered]
Now that Pages is also Standalone, does it mean that it can be installed "standalone next to a Joomla website" and that it would inject content in Joomla? Or does Standalone mean that it is really standalone, meaning:
- that you cannot rely for anything coming from Joomla
- and that you have then to create your own template in Pages itself, modules in Pages itself etc?

**Comment by johanjanssens** [comment, marked as answer]
In short, much better performance! There is a lot of overhead when you run Joomla to render a webpage:

- Joomla core is bootstrapped and loaded
- Plugins are loaded (and system plugins are all executed)
- Database connection is created
- PHP Session is created
- ...

Pages doesn't require any of the above to render a page. They are overhead. By running Pages standalone you are able to get the best performance possible. 

There is indeed a small bit of extra work since you have to create your own theme, use partials instead of modules and generate your own menu but that is all very easy to do. 

For the long version, keep reading.

## How a php web app works

PHP web applications like Joomla have a so-called entry point, this usually is an `index.php` file that is loaded by the webserver. Joomla contains 3 applications each which run `standalone`and each has it's own index.php file.

1. Site => /index.php
2. Administrator => /administrator/index.php
3. Installation => /installation/index.php

Pages can now be run in 2 setups:

### 1. Inside Joomla

When Pages is run inside Joomla, Joomla's index.php is loaded, Joomla boots our Joomlatools Framework, through a system plugin and the framework is responsible to bootup Pages.

### 2. Standalone

When Pages is run standalone, you need to add your own `index.php` file and tell your server to use it. In this `index.php` file you load the [Pages bootstrap script](https://github.com/joomlatools/joomlatools-pages/blob/master/resources/pages/bootstrapper.php), which does the same as the Joomla system plugin does, it boots the framework and it boots Pages.

Here is the code:

```php
Koowa::getInstance(array(
    'root_path'    => KOOWA_ROOT,
    'vendor_path'  => KOOWA_VENDOR,
    'cache'        => getenv('KOOWA_DEBUG') ? false : true,
));

//Bootstrap Framework
Koowa::getObject('object.bootstrapper')
    ->registerComponent('koowa', KOOWA_VENDOR.'/joomlatools/framework/code/libraries/joomlatools/component/koowa', 'koowa')
    ->registerComponent('pages', KOOWA_VENDOR.'/joomlatools/pages/code', 'koowa')
    ->bootstrap();

Koowa::getObject('event.publisher')->publishEvent('onAfterKoowaBootstrap');

//Dispatch Pages component
Koowa::getObject('com:pages.dispatcher.http')->dispatch();
```

## What standalone means for Pages

When you run Pages standalone, Pages is responsible to render the complete page. This means the following isn't available to you:

- Joomla templates
- Joomla langauges
- Joomla modules
- Joomla plugins

To create a site you need to create your own theme in Pages, not very hard. The result is much better performance and much more control on your end. You also don't need to bother setting up a database by default anymore. Pages doesn't require that to work, just PHP is all you need.

### Getting data out of Joomla

You can still integrate Pages with Joomla, but only on a data level only. There are two options here:

#### 1.  Connecting to the Joomla database

Instead of letting Joomla setup the database connection you would add the necessary info for this to your Pages config, much like you do in Joomla and then Pages can still use the Joomla extension and the different collection models to get data out of Joomla and render it on your site.

#### 2. Connecting to Joomla API

Another option is that you install Pages in the Joomla site you want to connect with and create a few pages to expose the Joomla data as a JSON webservice. This way you turn Joomla into a headless CMS. You can then use the webservice collection to connect to the endpoints in your Joomla site and get the data. 

Because you are using Pages on both ends, things will work seamlessly and you can make full use of all the smart caching that Pages offers.

**Reply by alexandreelise** [reply]
One more point. If we need to create our own templates and menus when using standalone version. Doesn't it imply that the standalone version is more for tech savvy people? (Let's be clear joomlatools audience is clearly developers) But am I right for pages standalone or completely wrong? @johanjanssens

**Reply by johanjanssens** [reply]
Not at all. Running Pages standalone or in Joomla requires the same level of knowledge, and I'm trying to make it as simple and joyful as possible. 

---

_Free Software is first and foremost about code, it's the source code that is open sourced. The GPL license gives you the freedom to use, study, and change the code. This also means that FOSS inherently requires a level of tech-savvy'ness._

---

Good FOSS software should have easy readable and understandable code, and should be simple to integrate.  Specifically for Pages you do not need to be a PHP developer to build things with Pages, you do require some notion of web development, html and css to get started. 

Take for example @woluweb his demo site: https://pages.joomlacustomfields.org/ Marc is not a PHP developer, he is tech-savvy enough to put the site together. 

There is not much difference between running Pages inside Joomla and running it standalone. You can run the exact same code both ways. Bootstrapping Pages standalone is really easy, with the lates release I got it down to one line of code:

````php
<?php
require getcwd().'/vendor/joomlatools/pages/resources/pages/bootstrapper.php'`
````
You put this line of code in an `index.php`file in your DOCUMENT_ROOT (same place as Joomla has it) and you are good to go. I'm still working to put together a boilerplate setup for Pages you can just copy past to get you started. 

### Menus

If you like to you can keep using Joomla as a decoupled CMS editing your menu items, and rendering them with Pages. This will make less sense in a standalone context, but nothing prevents you from doing that.

You can use the `ext:joomla.model.menus`to generate a list of menu items. All you for that to work is to add your database config to your Pages site `config.php` so that pages can connect to your database. 

There is a re-usable partial here that is copy-paste: https://github.com/joomlatools/joomlatools-pages/discussions/601 that can render a list of items coming from Joomla as well as Pages coming from the filesystem. 

### Templates

Joomla templates you cannot re-use, they are too tightly coupled to Joomla to be re-usable outside of it. A [Pages theme](https://github.com/joomlatools/joomlatools-pages/wiki/Themes) is very simple though, it's nothing more than a css file. 

I tried to make it as easy as possible to get started with the blog sample site:

- Code can be found here: https://github.com/joomlatools/joomlatools-pages/tree/master/contrib/sites/blog and
- Package for download: https://github.com/joomlatools/joomlatools-pages/releases/tag/v0.21.3 here, we build a separate package with each release that just contains the blog site

There are loads of html themes out there, you can re-use. For the blog sample site: https://github.com/joomlatools/joomlatools-pages/pull/671 we used [Blog 1](https://templates.digizu.co.uk/layouts/blog-1.html) from https://templates.digizu.co.uk/

**Reply by alexandreelise** [reply]
My question was not intented to offend you or anyone @johanjanssens . I am an experienced PHP and Joomla developer and I understand the tremedous power of standalone pages. I am myself starting to work with it on the new version of my pro website.
Thanks anyway for clarifying and expanding your point to be clear about what FOSS is. I don't think FOSS is only about code but since we are coders let's take it that way.

**Reply by johanjanssens** [reply]
No offense intended, edited my reply a little bit just wanted to provide some broader context, as this question comes up often. 

Just to further clarify, Joomlatools doesn't aim towards developers, our audience and customers are web integrators, they always have been. I would actually go as far to say that Joomla's audience are not developers either, developers would choose for solutions like Sympfony, Laravel or others. Joomla's audience is mostly web agencies and integrators, some have PHP skills, most don't.
